"Prototype an assistant" page with integrated editors, train and download buttons. (#6095)
RasaHQ/rasa: docs/user-guide/prototype-an-assistant.rst

Prototype an Assistant
======================

.. edit-link::

> Get started with Rasa Open source and learn how to create an assistant from scratch!

Reviewer: ```suggestion
Get started with Rasa Open Source and learn how to create an assistant from scratch!
```

> This page explains the basics of building an assistant with Rasa and shows the structure of a Rasa project.
> You can test it out right here without installing anything. You can also install Rasa and follow along in your command line.

Reviewer: ```suggestion
This page explains the basics of building an assistant with Rasa Open Source and shows the structure of a Rasa project.
You can test it out right here without installing anything. You can also :ref:`install Rasa Open Source <installation>` and follow along in your command line.
```

1. Define a basic user goal
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To prototype an assistant, start with a single `user goal <glossary>`_

Reviewer: ```suggestion
To prototype an assistant, start with a single :ref:`user goal <glossary>`
```

> that your assistant will handle. You can create a prototype using the user goal we have chosen
> "subscribing to a newsletter", or you can come up with your own. If you choose your own user goal,

Reviewer: ```suggestion
that your assistant will handle. You can create a prototype with the "subscribing to a newsletter" user goal we have chosen,
or you can come up with your own. If you choose your own user goal,
```

you can create a prototype that handles your goal by following the instructions to change each
section throughout the tutorial.

2. Create some NLU data
~~~~~~~~~~~~~~~~~~~~~~~

> `NLU <glossary>`_ data provides examples of what users might say to your assistant and what they mean by it.
> `Intent <glossary>`_  refers to what the user means with a specific message.

Reviewer: ```suggestion
:ref:`NLU <glossary>` data provides examples of what users might say to your assistant and what they mean by it.
An :ref:`intent <glossary>` refers to what the user means with a specific message.
```

You can delete the ``subscribe`` and ``inform`` intents if you’re not using them;
you can also add or change examples for any of the other intents.

.. code-editor::
    :language: yaml
    :id: nlu
    :height: 400
    :tracking_endpoint: https://trainer-service.prototyping.rasa.com/startPrototyping

    nlu:
    - intent: greet
      examples: |
           Hi
           Hey!
           Hallo
           Good day
           Good morning

    - intent: subscribe
      examples: |
           I want to get the newsletter
           Can you send me the newsletter?
           Can you sign me up for the newsletter?

    - intent: inform
      examples: |
           My email is example@example.com
           random@example.com
           Please send it to anything@example.com
           Email is something@example.com

:ref:`Read about best practices for NLU <training-data-format>`

3. Define a simple form
~~~~~~~~~~~~~~~~~~~~~~~

For most user goals, the bot will need to collect some information from the user to fulfill their request.
To do so, we define a `form <glossary>`_. You can change the name of the form to reflect your user goal.

> Add to or replace the "email" item in the list below to reflect the information the bot needs
> to collect for your user goal. Leave the "type" field the same for any items you add.

Reviewer: ```suggestion
Add to or replace the ``email`` item in the list below to reflect the information the bot needs
to collect for your user goal. Leave the ``type`` field the same for any items you add.
```

.. code-editor::
    :language: yaml
    :id: forms
    :height: 150
    :tracking_endpoint: https://trainer-service.prototyping.rasa.com/startPrototyping

>     forms:
>     - newsletter_form:
>        email:
>        - type: from_text

Reviewer: ```suggestion
    forms:
    - newsletter_form:
        email:
        - type: from_text
```
```suggestion
    forms:
    - newsletter_form:
        email:
        - type: from_text
```

Reviewer: slash maybe in general we need to pick an indentation level and stick to it (there was 1, 2, and 4 i believe)

4. Write a rule
~~~~~~~~~~~~~~~

> Rules are conversation snippets that should always precede in exactly the same way. The ellipsis in the

Reviewer: ```suggestion
Rules are conversation snippets that should always proceed in exactly the same way. The ellipsis in the
```

first step of the rule tells the bot that regardless of what precedes, the next steps should always
follow one another. Update the rule below for running the form you defined above.

.. code-editor::
    :language: yaml
    :id: rules
    :height: 200
    :tracking_endpoint: https://trainer-service.prototyping.rasa.com/startPrototyping

    rules:
     - rule: activate subscribe form
       steps:
       - ...
       - intent: subscribe
       - action: newsletter_form
       - form: newsletter_form

5. Write some stories
~~~~~~~~~~~~~~~~~~~~~~~

Stories are example conversations of how your assistant should handle a user’s intent in context.

> The first stories you write should follow the `happy path <glossary>` for your user goal.

Reviewer: ```suggestion
The first stories you write should follow the :ref:`happy path <glossary>` for your user goal.
```

> A story contains one or more blocks of (user) intent and (bot) `actions <glossary>`_ or `responses <glossary>`_.

Reviewer: ```suggestion
A story contains one or more blocks of (user) intent and (bot) :ref:`actions <glossary>` or :ref:`responses <glossary>`.
```

.. code-editor::
    :language: yaml
    :id: stories
    :height: 250
    :tracking_endpoint: https://trainer-service.prototyping.rasa.com/startPrototyping

    stories:
     - story: greet and subscribe
       steps:
       - intent: greet
       - action: utter_greet
       - intent: subscribe
       - action: newsletter_form
       - form: newsletter_form

:ref:`Read about best practices for writing stories <stories>`

6. Edit responses
~~~~~~~~~~~~~~~~~
To give your bot messages to respond to the user with, you need to define responses.
You can specify one or more text options for each response. If there are multiple, one of the options
will be chosen at random whenever that response is predicted.

You can add or change text for any of the responses below. If you’re using your own user goal,
replace the last three responses with the response you used in your stories above.

.. code-editor::
    :language: yaml
    :id: responses
    :height: 300
    :tracking_endpoint: https://trainer-service.prototyping.rasa.com/startPrototyping

    responses:
       utter_greet:
           - text: |
               Hallo! How can I help you?
           - text: |

>                hi!

Reviewer: ```suggestion
               Hi!
```

       utter_ask_email:
           - text: |
               What is your email address?
       utter_subscribed:
           - text: |
               I’ve subscribed {email} to the newsletter!
           - text: |
               You’ve been subscribed, the newsletter will be sent to {email}.

.. note::

    For this prototype, we have only defined responses, meaning the only thing the assistant does is respond with a
    predefined message. Custom actions, however, can be defined to do whatever you’d like.
    For example, for the user goal of subscribing to a newsletter, you could create a custom action

>     that adds the user’s email to a database. You can see an example of this in `Sara's action code <action code>`_.

Reviewer: This needs a proper link

7. Train and run
~~~~~~~~~~~~~~~~

Rasa has a command line interface that allows you to train and run your bot from a terminal.
To train your bot on the NLU data, stories and responses you’ve just defined, run ``rasa train`` using
the button below:

.. train-button::
    :endpoint: https://trainer-service.prototyping.rasa.com/trainings

>     :method: POST

Reviewer: This takes a pretty long time -- where is the config stored? can we reduce the epochs?

Reviewer: Ah i guess we want to keep it the standard for when it's downloaded though

Author: I guess it's fine now

Once your model has finished training, you can talk to your assistant:

.. chat-block::

8. What's next?
~~~~~~~~~~~~~~~

You can download this project and build on it to create a more advanced assistant.
In your downloaded project, you’ll notice several files that were configured for you that you didn’t edit on this page.

> To learn more about configs, domains and actions, refer to the advanced tutorials.

Reviewer: link to tutorials, or to their docs pages

.. note::

    For this prototype, we have only defined responses, meaning the only thing the assistant does is respond with a
    predefined message. Custom actions, however, can be defined to do whatever you’d like.
    For example, for the user goal of subscribing to a newsletter, you could create a custom action
    that adds the user’s email to a database. You can see an example of this
    in `Sara's action code <https://github.com/RasaHQ/rasa-demo/blob/master/actions/actions.py>`_.

7. Train and run
~~~~~~~~~~~~~~~~

Rasa has a command line interface that allows you to train and run your bot from a terminal.
To train your bot on the NLU data, stories and responses you’ve just defined, run ``rasa train`` using
the button below:

.. train-button::
    :endpoint: https://trainer-service.prototyping.rasa.com/trainings
    :method: POST

Once your model has finished training, you can talk to your assistant:

.. chat-block::

8. What's next?
~~~~~~~~~~~~~~~

You can download this project and build on it to create a more advanced assistant.
In your downloaded project, you’ll notice several files that were configured for you that you didn’t edit on this page.

> To learn more about configs, domains and actions, refer to the advanced tutorials.

Reviewer: ```suggestion
Check out other docs pages to learn more about :ref:`domains`, :ref:`actions`,
and your config's :ref:`pipeline <choosing-a-pipeline>` and :ref:`policies`.
```
